Some algorithm benchmarks written in Java.

zafer-esen opened the pull request:
<!--
  Please describe your PR as usual.

  For submission of new verification tasks,
  keep the following checklist and make sure that all items are fullfilled.
  For other PRs, just remove it.
-->

- [ ] programs added to new and [appropriately named](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#directory-structure-and-names) directory
- [ ] license present and [acceptable](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#license) (either in separate file or as comment at beginning of program)
- [ ] [contributed-by](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#origin-description-and-attribution) present (either in README file or as comment at beginning of program)

- [ ] programs added to a `.set` file of an existing category, or new sub-category established (if justified)
- [ ] intended property matches the corresponding `.prp` file
- [ ] programs and expected answer added to a `.yml` file according to [task definitions](https://github.com/sosy-lab/sv-benchmarks#task-definitions)

<!-- For C programs: -->
- [ ] [architecture](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#architecture) (32 bit vs. 64 bit) matches the corresponding `.cfg` file
- [ ] original sources present
- [ ] [preprocessed](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#preprocessing) files present
- [ ] preprocessed files generated with correct architecture
- [ ] [Makefile](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#compile-checks) added with correct content and without overly broad suppression of warnings

mmuesly commented on java/algorithms/SortedListInsert-FunUnsat01/Main.java:
```diff
@@ -0,0 +1,87 @@
+    for (int i = 0; i < N; i++)
+      list.insert(Verifier.nondetInt());
+
+    assert(list.x < list.next.x); // the values can be equal
```
Same as above.

mmuesly commented on java/algorithms/SortedListInsert-MemSat01/Main.java:
```diff
@@ -0,0 +1,84 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
```
From my point of view, this example is missing an assert statement. What is the property that should be checked here?

mmuesly commented on java/algorithms/SortedListInsert-MemUnsat01/Main.java:
```diff
@@ -0,0 +1,84 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
```
From my point of view, this example is missing an assert statement. What is the property that should be checked here?

mmuesly commented on java/algorithms/Trie-FunSat01/Main.java:
```diff
@@ -0,0 +1,286 @@
+      int val = Verifier.nondetInt();
+      st.put(test, val);
+      assert (st.get(test) == val);
+    } catch(Exception e) { assert false; }
```
After reading through the code, I am not sure, whether this catch is ever supposed to be triggered.
Otherwise, might the try-catch block be removed?

zafer-esen replied:
Yes, I've removed it for the benchmarks checking functional safety. For the benchmarks checking memory safety, I guess the answer will depend on the conclusion of [this conversaton](https://github.com/sosy-lab/sv-benchmarks/pull/823#discussion_r338071487).

mmuesly commented on java/algorithms/Trie-FunUnsat01/Main.java:
```diff
@@ -0,0 +1,286 @@
+ * Original license follows.
+ */
+
+/******************************************************************************
```
What is the license here? I don't see any.

zafer-esen replied:
Added license.

mmuesly commented on java/algorithms/Trie-MemSat01/Main.java:
```diff
@@ -0,0 +1,285 @@
+ * Original license follows.
+ */
+
+/******************************************************************************
```
What is the license here? I don't see any.

mmuesly commented on java/algorithms/Trie-MemUnsat01/Main.java:
```diff
@@ -0,0 +1,285 @@
+ * Original license follows.
+ */
+
+/******************************************************************************
```
What is the license here? I don't see any.

zafer-esen replied:
Added license.

mmuesly commented on java/algorithms/Tsp-MemSat01/Main.java:
```diff
@@ -0,0 +1,123 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
```
From my point of view, this file is missing an assert statement. What is the property, that should be proved here?

mmuesly commented on java/algorithms/Tsp-MemUnsat01/Main.java:
```diff
@@ -0,0 +1,123 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
```
From my point of view, this file is missing an assert statement. What is the property, that should be proved here?

mmuesly commented on java/algorithms/MergeSortIterative-FunUnsat01/Main.java:
```diff
@@ -0,0 +1,141 @@
+ */
+
+// IterativeMergeSort.java
+// By David Kosbie
```
I think this is not a proper license. Might you clarify it?

mmuesly commented on java/algorithms/MergeSortIterative-FunSat01/Main.java:
```diff
@@ -0,0 +1,127 @@
+
+// IterativeMergeSort.java
+// By David Kosbie
+
```
I think this is not a proper license. Might you clarify it?

mmuesly commented on java/algorithms/BellmanFord-FunSat01/Main.java:
```diff
@@ -0,0 +1,109 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
+
```
General comment: I think all files need a clear comment under which license the modifications are published.
Further, there should be a clear indication whether the files is modified or not and how is the author of the modifications.

zafer-esen commented on java/algorithms/BellmanFord-MemSat01/Main.java:
```diff
@@ -0,0 +1,104 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
```
The benchmarks with the extension MemSat and MemUnsat are all without explicit assertions, and the property being checked is that there are no thrown standard Java exceptions. As you've mentioned in another comment we have assumed that the verifier should check the reachability of these standard exceptions. 

If this is not suitable, maybe we can wrap the potentially unsafe code with try-catch blocks, with `assert false;` in the catch statement?

zafer-esen commented on java/algorithms/BellmanFord-FunUnsat01/.#Main.java:
```diff
@@ -0,0 +1 @@
+zafer@it-ll-zafes790.1831:1570607289
```
No, removing it.

zafer-esen commented on java/algorithms/BinaryTreeSearch-FunSat01/Main.java:
```diff
@@ -0,0 +1,135 @@
+    for (int i = 1; i < N; i++){
+      int n = Verifier.nondetInt();
+      if(n >= 0) // only insert positive numbers
+        b.insert(n);
```
Added curly brackets.

zafer-esen commented on java/algorithms/BinaryTreeSearch-FunSat01/Main.java:
```diff
@@ -0,0 +1,135 @@
+        b.insert(n);
+    }
+
+    assert(b.search(-1) == false); // didn't put in any negative numbers
```
Not sure if I understood this. Didn't put in any negative numbers, so the assertion is making sure that `b.search(-1)` evaluates to `false`.

mmuesly replied:
Sure, but what is about -2 ... Integer.MIN_VALUE ?
The assert doesn't check that they are neither in the tree.

mmuesly replied:
Looks good to me.

zafer-esen commented on java/algorithms/BinaryTreeSearch-FunSat01.yml:
```diff
@@ -0,0 +1,8 @@
+  - BinaryTreeSearch-FunSat01/
+properties:
+  - property_file: ../properties/assert.prp
+    expected_verdict: True
```
Fixed.

zafer-esen commented on java/algorithms/BinaryTreeSearch-FunUnsat01.yml:
```diff
@@ -0,0 +1,8 @@
+  - BinaryTreeSearch-FunUnsat01/
+properties:
+  - property_file: ../properties/assert.prp
+    expected_verdict: False
```
Fixed.

zafer-esen commented on java/algorithms/BinaryTreeSearch-MemSat01.yml:
```diff
@@ -0,0 +1,8 @@
+  - BinaryTreeSearch-MemSat01/
+properties:
+  - property_file: ../properties/assert.prp
+    expected_verdict: True
```
Fixed.

zafer-esen commented on java/algorithms/BinaryTreeSearch-MemUnsat01.yml:
```diff
@@ -0,0 +1,8 @@
+  - BinaryTreeSearch-MemUnsat01/
+properties:
+  - property_file: ../properties/assert.prp
+    expected_verdict: False
```
Fixed.

zafer-esen commented on java/algorithms/BinaryTreeSearch-MemUnsat02.yml:
```diff
@@ -0,0 +1,8 @@
+  - BinaryTreeSearch-MemUnsat02/
+properties:
+  - property_file: ../properties/assert.prp
+    expected_verdict: False
```
Fixed.

zafer-esen commented on java/algorithms/BellmanFord-FunSat01/Main.java:
```diff
@@ -0,0 +1,109 @@
+    for (int i = 0; i < V; i++) {
+      for (int j = 0; j < V; j++) {
+        if (i == j) continue;
+        D[i*V+j] =  Verifier.nondetInt();
```
I am not sure how we can strengthen the assertions without making them almost impossible to verify. Maybe we can add two assumptions? 
- assume that V is at most 1e6, which should still be almost impossible to verify using bounded methods and cause no overflows in the calculations
- assume the same with the initialization of `D` at line 98

mmuesly replied:
I think the assumptions might be fine for now. Nevertheless, being able to verify such functional correctness properties and find a way to express them might be an interesting challenge in the long term.

zafer-esen commented on java/algorithms/BellmanFord-FunSat01/Main.java:
```diff
@@ -0,0 +1,109 @@
+ */
+public class Main {
+
+  static final int INFINITY = 1000000;
```
There is no special reason, this was carried over from the original benchmark. Changing it as
`INFINITY = Integer.MAX_VALUE` makes sense to me, will update it this way if you agree.

mmuesly replied:
Sounds good to me.

zafer-esen commented on java/algorithms/BellmanFord-FunSat01/Main.java:
```diff
@@ -0,0 +1,109 @@
+    final int V = Verifier.nondetInt();
+    if (V <= 0) return; // change with Verifier.assume?
+
+    final int D[] = new int[V*V];
```
I guess the change `INFINITY = Integer.MAX_VALUE` should resolve this:
> nit-pick:
`INFINITY < V` might occur using `Verifier.nondetInt()`

Adding the assumptions I've mentioned in another comment should resolve the overflow issue I think: 
- assume that V is at most 1e6, which should still be almost impossible to verify using bounded methods and cause no overflows in the calculations
- assume the same with the initialization of D at line 98

> Specifying the reachability of certain Standard Java-Exceptions might be one option report this overflow or we might add an assert statement that ensures no overflow before the int.

We have assumed that the verifier should check the reachability of these standard exceptions. If this is not suitable, maybe we can wrap the potentially unsafe code with try-catch blocks, with assert false; in the catch statement?

mmuesly replied:
It is my understanding that currently only explicit assert violations are checked in this category according to the property file. Using a try-catch blog sounds legit to me. But I might be wrong. 
Otherwise it might be possible to define a new property for checking the standard errors on a program and use such a property along with this file. I am not sure, what would be the better approach.

Eventually @peterschrammel or @dbeyer might give some guidance for the better solution here.

zafer-esen commented on java/algorithms/InsertionSort-FunSat01/Main.java:
```diff
@@ -0,0 +1,83 @@
+        a[j + 1] = a[j];
+        j--;
+      }
+	    count += N;
```
It's not really being used anywhere, it was in the original code to keep track of the array length. It shouldn't affect the verdict, but I've removed it anyway since it's redundant.

zafer-esen commented on java/algorithms/RedBlackTree-FunUnsat01/rbtree/Node.java:
```diff
@@ -0,0 +1,112 @@
+/**
```
It is modified in RedBlackTree-MemUnsat01 to introduce an artificial error (in RedBlackTree.java:358). The other three RedBlackTree benchmarks have it in common; but, I'm not sure if it's a good idea to have a joint directory only for those three.

mmuesly replied:
okay

zafer-esen commented on java/algorithms/SortedListInsert-FunSat01/Main.java:
```diff
@@ -0,0 +1,87 @@
+    for (int i = 0; i < N; i++)
+      list.insert(Verifier.nondetInt());
+
+    assert(list.x <= list.next.x);
```
It's a good point. Maybe we can add multiple versions with assertions of different strength? Some verifiers might be able to prove the weaker assertion that the first two elements are sorted, but not the whole list.  This actually applies to all sorting algorithm benchmarks.

mmuesly replied:
Sounds like a good proposal to me.

mmuesly commented on java/algorithms/RedBlackTree-FunUnsat01/Main.java:
```diff
@@ -0,0 +1,69 @@
+
+public class Main {
+  public static void main(String[] args) {
+    int N = Verifier.nondetInt();
```
I see the point here. Hence, I would support this proposal.

dbeyer commented:
@zafer-esen Could you please complete this pull request such that it can be merged?
People would like to train on the programs.

zafer-esen commented:
> @zafer-esen Could you please complete this pull request such that it can be merged?
> People would like to train on the programs.

Hi @dbeyer, I have just pushed a new commit covering most of the requested changes. 

I have not done anything regarding the issue [in this comment](https://github.com/sosy-lab/sv-benchmarks/pull/823#discussion_r338071487). Depending on your feedback, I can add the try-catch blocks in the memory safety benchmarks in this pull request or separately.

Addiiton of more benchmarks with different assertion strength levels proposed [in this comment](https://github.com/sosy-lab/sv-benchmarks/pull/823#discussion_r338081500) is also not in the latest commit. I can maybe create another pull request for those?

dbeyer commented:
@zafer-esen I would say yes to both comments (yes, in the competition we check (currently) only for asserts that are reached and violated, yes, please add the variants, but in a new pull request).

zafer-esen commented:
@dbeyer and @mmuesly, thank you both for your comments. I have added the assertions via try-catch blocks to all the benchmarks which lacked these, in the latest commit. 

Regarding [this comment](https://github.com/sosy-lab/sv-benchmarks/pull/823#discussion_r337986760), I have not deleted the MergeSortIterative* benchmarks as I have contacted the original author David Kosbie (<koz@cmu.edu>), and he kindly granted permission to freely use this code. I have added this info to the headers of the relevant benchmarks, if this will suffice. 

I will create another pull request for the sorting algorithm benchmaks, with assertions which test stronger properties.

dbeyer commented:
@mmuesly Could you please have another look? Is this pull request ready to be merged from your point of view?

PhilippWendler commented:
> Does CI make sure these test cases are compilable? I haven't checked this myself, but we might fix this along the way, if problems arise.

CI checks test cases if they are added to `compile.xml`, which this PR [does](https://github.com/sosy-lab/sv-benchmarks/pull/823/files?file-filters%5B%5D=.xml#diff-921d218e88ae8f5e0b43d681b9dd7efe). This CI check just runs `javac` with all `.java` files below any of the directories in `input_files`.
